### PR TITLE
cnvkit module

### PR DESCRIPTION
**Pull request by kaurravneet4123**
PR for cnvkit module file.

### CODE REVIEW

**Review by drpatelh — COMMENTED** [review]
Thanks @kaurravneet4123 ! @MaxUlysse fancy reviewing this one? Seeing as it's something that is likely to go into Sarek too?

File: software/cnvkit/main.nf
Comment (MaxUlysse) [inline software/cnvkit/main.nf:52]:
```suggestion
          cnvkit.py version > ${software}.version.txt

```

File: software/cnvkit/main.nf
Comment (MaxUlysse) [inline software/cnvkit/main.nf:62]:
```suggestion
          cnvkit.py version > ${software}.version.txt
```

File: software/cnvkit/main.nf
Comment (MaxUlysse) [inline software/cnvkit/main.nf:8]:
```suggestion
```

File: software/cnvkit/main.nf
Comment (MaxUlysse) [inline software/cnvkit/main.nf:49]:
```suggestion
          --fasta $fasta \\
```

**Comment by MaxUlysse** [comment]
The `version` command can be used to get the version. cf suggestions.
I think the `--method wgs` and `--output-reference reference.cnn --output-dir output` should go out of the process, and go in `options.args`.
Only thing I can see missing would be a `--targets ${target_bed}` option.
I'm not sure about the `meta.test_with_normal`
Are you already using something similar in a pipeline?
I would rather separate the 2 usages in 2 processes that would take different input.
But actually not sure what the consensus is on that matter, so I'll ping @drpatelh for some wisdom on that.

**Comment by kaurravneet4123** [comment]
Hi Maxime and Harshil,
Many thanks for your comments. 
Sorry I couldn't reply earlier. I had a few urgent things to work on.
I will work on the comments tonight and will get back to you soon.
Kind regards,
Rav


signature   Ravneet 

    On Wednesday, February 17, 2021, 02:48:57 PM GMT, Maxime Garcia <notifications@github.com> wrote:  
 
 


The version command can be used to get the version. cf suggestions.
I think the --method wgs and --output-reference reference.cnn --output-dir output should go out of the process, and go in options.args.
Only thing I can see missing would be a --targets ${target_bed} option.
I'm not sure about the meta.test_with_normal
Are you already using something similar in a pipeline?
I would rather separate the 2 usages in 2 processes that would take different input.
But actually not sure what the consensus is on that matter, so I'll ping @drpatelh for some wisdom on that.

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf:62]: change noted

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf:8]: change done.

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf:49]: change done!

Comment (kaurravneet4123) [reply software/cnvkit/main.nf:49]: > The `version` command can be used to get the version. cf suggestions.
> I think the `--method wgs` and `--output-reference reference.cnn --output-dir output` should go out of the process, and go in `options.args`.
> Only thing I can see missing would be a `--targets ${target_bed}` option.
> I'm not sure about the `meta.test_with_normal`
> Are you already using something similar in a pipeline?
> I would rather separate the 2 usages in 2 processes that would take different input.
> But actually not sure what the consensus is on that matter, so I'll ping @drpatelh for some wisdom on that.




Hi Maxime,
> I have deleted `--method wgs` and `--output-reference reference.cnn --output-dir output`  from the process. Where should it go then? in the test.yml?
> --targets is missing because currently this module file is for WGS data which does not require target.bed. It is a target panel data that requires --targets. Once the module file is successful for the WGS data, I will add --target for the target panel also in a different process

**Comment by MaxUlysse** [comment]
@kaurravneet4123 
> I have deleted --method wgs and --output-reference reference.cnn --output-dir output from the process. Where should it go then? in the test.yml?
In the `main.nf` used to do the test actually. you should be able to provide that using `addParams( options: ['args': '--method wgs --output-reference reference.cnn --output-dir output'] )` after the include statement.

> --targets is missing because currently this module file is for WGS data which does not require target.bed. It is a target panel data that requires --targets. Once the module file is successful for the WGS data, I will add --target for the target panel also in a different process
I don't think we need a different process for that actually.
Using an optional input would work well in this case.
I'm even wondering if we could also use optional input for the `--annotate $annotationfile` part, and maybe optional input could be a better solution to the `meta.test_with_normal` issue.

**Comment by kaurravneet4123** [comment]
Thanks Maxime. I will edit the main.nf by using addParams.
I agree- we can use optional input for the --targets and for the --annotate arguments as they are not required in all the cases. 
For the next 2 weeks, I am planning to devote myself fully to the cnvkit module work with the aim to make it fully functional by 5th March - both for the sarek pipeline and for my workplace. Do you think it is a realistic goal? If you are happy, we both can aim together to finish this by 5th March. What do you think?
I can then start working on a new module during the hackathon. I am inclined towards working on 'manta' . 





signature   Ravneet 

    On Friday, February 19, 2021, 07:46:50 AM GMT, Maxime Garcia <notifications@github.com> wrote:  
 
 


@kaurravneet4123


I have deleted --method wgs and --output-reference reference.cnn --output-dir output from the process. Where should it go then? in the test.yml?


In the main.nf used to do the test actually. you should be able to provide that using addParams( options: ['args': '--method wgs --output-reference reference.cnn --output-dir output'] ) after the include statement.


--targets is missing because currently this module file is for WGS data which does not require target.bed. It is a target panel data that requires --targets. Once the module file is successful for the WGS data, I will add --target for the target panel also in a different process


I don't think we need a different process for that actually.
Using an optional input would work well in this case.
I'm even wondering if we could also use optional input for the --annotate $annotationfile part, and maybe optional input could be a better solution to the meta.test_with_normal issue.

**Comment by MaxUlysse** [comment]
> Thanks Maxime. I will edit the main.nf by using addParams.
> I agree- we can use optional input for the --targets and for the --annotate arguments as they are not required in all the cases.
> For the next 2 weeks, I am planning to devote myself fully to the cnvkit module work with the aim to make it fully functional by 5th March - both for the sarek pipeline and for my workplace. Do you think it is a realistic goal? If you are happy, we both can aim together to finish this by 5th March. What do you think?

More than realistic, I even think that March 5th is a bit too pessimistic given your coding skills.
But better to plan more time and have pleasant surprises than the other way around.

>  I can then start working on a new module during the hackathon. I am inclined towards working on 'manta' 

Why not, it'll definitively be helpful.

**Comment by kaurravneet4123** [comment]
Thanks!

signature   Ravneet 

    On Monday, February 22, 2021, 10:16:56 AM GMT, Maxime Garcia <notifications@github.com> wrote:  
 
 



Thanks Maxime. I will edit the main.nf by using addParams.
I agree- we can use optional input for the --targets and for the --annotate arguments as they are not required in all the cases.
For the next 2 weeks, I am planning to devote myself fully to the cnvkit module work with the aim to make it fully functional by 5th March - both for the sarek pipeline and for my workplace. Do you think it is a realistic goal? If you are happy, we both can aim together to finish this by 5th March. What do you think?


More than realistic, I even think that March 5th is a bit too pessimistic given your coding skills.
But better to plan more time and have pleasant surprises than the other way around.


I can then start working on a new module during the hackathon. I am inclined towards working on 'manta'


Why not, it'll definitively be helpful.

**Comment by kaurravneet4123** [comment]
Hi Maxime,
which directory would be the launch directory in my case?
and what should be the value for meta.id? Is it related to the name of the input or output files?
signature   Ravneet 

    On Monday, February 22, 2021, 10:16:56 AM GMT, Maxime Garcia <notifications@github.com> wrote:  
 
 



Thanks Maxime. I will edit the main.nf by using addParams.
I agree- we can use optional input for the --targets and for the --annotate arguments as they are not required in all the cases.
For the next 2 weeks, I am planning to devote myself fully to the cnvkit module work with the aim to make it fully functional by 5th March - both for the sarek pipeline and for my workplace. Do you think it is a realistic goal? If you are happy, we both can aim together to finish this by 5th March. What do you think?


More than realistic, I even think that March 5th is a bit too pessimistic given your coding skills.
But better to plan more time and have pleasant surprises than the other way around.


I can then start working on a new module during the hackathon. I am inclined towards working on 'manta'


Why not, it'll definitively be helpful.

File: tests/software/cnvkit/main.nf
Comment (MaxUlysse) [inline tests/software/cnvkit/main.nf:6]: You can't include twice the same module if they have the same name.
You should be able to use `as` to rename the module, similar to what we're doing in the TABIX_TABIX module for example:
https://github.com/MaxUlysse/nf-core_modules/blob/01c0d335ca57b1b9298a7acd643116917953804a/tests/software/tabix/tabix/main.nf#L5-L7

File: software/cnvkit/main.nf
Comment (MaxUlysse) [reply software/cnvkit/main.nf:8]: I don't think you actually pushed any of the change you did related to this comment

**Comment by MaxUlysse** [comment]
> Hi Maxime, which directory would be the launch directory in my case?

I'm guessing you ask that related to the test in the `tests/software/cnvkit/main.nf` script.
And it looks fine to me, except that requesting the same module twice with the same name won't work.
Check out the comment I made about it.

> and what should be the value for meta.id?

I would just use `input = [ [ id:'test' ], // meta map` for the meta map in your test and get rid of the `meta.with_normal`.

> Is it related to the name of the input or output files?

the `meta.id` will be related to the output if you use the prefix define here: https://github.com/nf-core/modules/pull/173/files#diff-63daa746d582c734b9136b757431931bce33efcfe973a400126dd8af5699065cR43

**Comment by kaurravneet4123** [comment]
Thanks
signature   Ravneet 

    On Monday, February 22, 2021, 02:54:39 PM GMT, Maxime Garcia <notifications@github.com> wrote:  
 
 



Hi Maxime, which directory would be the launch directory in my case?


I'm guessing you ask that related to the test in the tests/software/cnvkit/main.nf script.
And it looks fine to me, except that requesting the same module twice with the same name won't work.
Check out the comment I made about it.


and what should be the value for meta.id?


I would just use input = [ [ id:'test' ], // meta map for the meta map in your test and get rid of the meta.with_normal.


Is it related to the name of the input or output files?


the meta.id will be related to the output if you use the prefix define here: https://github.com/nf-core/modules/pull/173/files#diff-63daa746d582c734b9136b757431931bce33efcfe973a400126dd8af5699065cR43

File: tests/software/cnvkit/main.nf
Comment (kaurravneet4123) [reply tests/software/cnvkit/main.nf:6]: ok. Thanks

Comment (kaurravneet4123) [reply tests/software/cnvkit/main.nf:6]: Maxime, do I need to add my institute's profile in the config file to use or test this module at my work place?

Comment (MaxUlysse) [reply tests/software/cnvkit/main.nf:6]: The whole explanation for CI is here: https://github.com/nf-core/modules#testing
You should be able to test on your local computer, no need to do that on your institute HPC.
But it's actually a good question @drpatelh @KevinMenden Is it possible to test locally using an institute profile?

**Comment by kaurravneet4123** [comment]
Hi Maxime,

How can I get rid of the following error (EditorConfig error screenshot attached):
Screenshot (278)

**Comment by kaurravneet4123** [comment]
I have seen strange lines appearing in the software/cnvkit/main.nf. Attached is the screen shot below. Is it normal for them to appear, e.g. line 46, 50-56, 62, 65-71?
Screenshot (279)

**Comment by MaxUlysse** [comment]
> I have seen strange lines appearing in the software/cnvkit/main.nf. Attached is the screen shot below. Is it normal for them to appear, e.g. line 46, 50-56, 62, 65-71?
> Screenshot (279)

This would be caused by some conflicts that were not resolved when merging code.
You can safely remove them, and the duplicated code if any

**Comment by MaxUlysse** [comment]
> Hi Maxime,
> 
> How can I get rid of the following error (EditorConfig error screenshot attached):
> Screenshot (278)

Given the error, I would say that you might have lines ending with an extra space, or some empty lines.
I'll try to have a look at your branch to see where it might be coming from.

File: tests/software/cnvkit/main.nf
Comment (KevinMenden) [reply tests/software/cnvkit/main.nf]: Sorry I overlooked that mention somehow .... not really sure about running locally with a different config profile.
There shouldn't be any extra configs for the module necessary, I think as long as you have one for the pipeline (which you'll need anyway), you're fine (not really adding much knowledge here I now 😄 )

Comment (kaurravneet4123) [reply tests/software/cnvkit/main.nf]: thanks Kevin

**Comment by kaurravneet4123** [comment]
Kindly if you can have a look at the addParams line in the tests/software/main.nf (highlighted in the screenshot below).  Are the map values correct? ( or do they need to be like --annotate $annotatefile)

Screenshot (281)

**Comment by kaurravneet4123** [comment]
Hi Maxime,

I am getting the following input channel error with pytest which is confusing me (attached is the screenshot): 

Screenshot (283)

Please if you can help me in resolving this.

PS: I have made a lot of changes in different files by reducing the workflows from 2 to 1. Single workflow might suit us better. What do you think?

**Comment by kaurravneet4123** [comment]
Morning @MaxUlysse ,

Pytest is successful :-),  please see the screenshots below.  Please if you can guide me for further steps.

Screenshot (290)

Screenshot (289)

**Comment by kaurravneet4123** [comment]
But on the https://github.com/nf-core/modules/pull/173, it shows that pytest-workflow is not successful!  This is strange!

Screenshot (291)

**Comment by kaurravneet4123** [comment]
@MaxUlysse: I think the pytests are failing at https://github.com/nf-core/modules/pull/173 because of the path of the main.nf in the tests/software/cnvkit/main.nf (screenshot attached). Is it supposed to be something else? 

Screenshot (292)

**Comment by MaxUlysse** [comment]
> @MaxUlysse: I think the pytests are failing at #173 because of the path of the main.nf in the tests/software/cnvkit/main.nf (screenshot attached). Is it supposed to be something else?
> 
> Screenshot (292)

Yes, you have an extra `../`, it should be only `'../../../software/cnvkit/main.nf`

**Comment by kaurravneet4123** [comment]
> > @MaxUlysse: I think the pytests are failing at #173 because of the path of the main.nf in the tests/software/cnvkit/main.nf (screenshot attached). Is it supposed to be something else?
> > Screenshot (292)
> 
> Yes, you have an extra `../`, it should be only `'../../../software/cnvkit/main.nf`

Thanks @MaxUlysse. I have made the change and pushed. And pytest is successful- at last :-)  . What is the next step- I think, I should add --targets arguments for the target panel data. It might need options.args2. Any more suggestions about themodule?

Screenshot (294)

File: software/cnvkit/main.nf
Comment (drpatelh) [inline software/cnvkit/main.nf:23]:
```suggestion
    tuple val(meta), path(tumourbam), path(normalbam)

```

File: software/cnvkit/main.nf
Comment (drpatelh) [inline software/cnvkit/main.nf:41]:
```suggestion
    cnvkit.py batch \\
        $tumourbam \\
        --normal $normalbam \\
        --fasta $fasta \\
        --annotate $annotationfile \\
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Mostly looks good @kaurravneet4123 ! You will need to change the `input` channels to explicitly define the BAM files.

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf]: change done

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf]: change done

**Comment by kaurravneet4123** [comment]
Hi @drpatelh. I have edited the reference fasta files in the tests/data/fasta/human/ dir. Earlier I was using hg19. Now I have changed it to GRCh38, as the input bam files were created using GRCh38 reference genome.

**Comment by kaurravneet4123** [comment]
Hi @drpatelh, @KevinMenden and @MaxUlysse, the sarscov2 test-data is working with the cnvkit module, after editing the bed file. Pytest with conda is successful in my local repo (see the attached screenshot) and I have pushed the changes. But in the PR, conda pytest is not appearing as successful! Why there is this contradiction?

Screenshot (334)

Screenshot (335)

**Comment by KevinMenden** [comment]
Hi @kaurravneet4123 , sorry for the late reply.

I checked the test logs and the reason is that your file paths seems to be wrong:
`java.nio.file.NoSuchFileException: /home/runner/pytest_workflow_qdlu29_6/cnvkit/tests/data/fasta/sarscov2/GCA_011545545.1_ASM1154554v1_genomic.fna`

The test data is in `tests/data/genomic/sarscov2/fasta/` and the name is also different, this might have changed while you were working on your PR. Sorry! You should probably pull the latest changes to be up to date with the current repo. Then it should be just a matter of switching the file paths.

Awesome that it runs through with the Sarscov2 data :muscle: :+1: 
I will try to have a closer look at this today or tomorrow, but for you you should try pulling the latest changes from `nf-core/modules` and then adapting your file paths.

**Comment by kaurravneet4123** [comment]
> Hi @kaurravneet4123 , sorry for the late reply.
> 
> I checked the test logs and the reason is that your file paths seems to be wrong:
> `java.nio.file.NoSuchFileException: /home/runner/pytest_workflow_qdlu29_6/cnvkit/tests/data/fasta/sarscov2/GCA_011545545.1_ASM1154554v1_genomic.fna`
> 
> The test data is in `tests/data/genomic/sarscov2/fasta/` and the name is also different, this might have changed while you were working on your PR. Sorry! You should probably pull the latest changes to be up to date with the current repo. Then it should be just a matter of switching the file paths.
> 
> Awesome that it runs through with the Sarscov2 data 💪 👍
> I will try to have a closer look at this today or tomorrow, but for you you should try pulling the latest changes from `nf-core/modules` and then adapting your file paths.


No problem @KevinMenden. Thanks for getting back to me. I will pull the latest changes and change the file paths :-)

**Comment by kaurravneet4123** [comment]
Hi @KevinMenden, I am trying to resolve the issue with conflicting files- but it is not resolving! I am not sure why the 2 files (.github/filters.yml and tests/software/TOOL/SUBTOOL/test.yml) are conflicting.  As a result, I am not able to merge the pull request. Please if you can guide me.

**Comment by kaurravneet4123** [comment]
> Hi @KevinMenden, I am trying to resolve the issue with conflicting files- but it is not resolving! I am not sure why the 2 files (.github/filters.yml and tests/software/TOOL/SUBTOOL/test.yml) are conflicting. As a result, I am not able to merge the pull request. Please if you can guide me.


Morning @Kevin, the issue is resolved. I should be able to progress further now.

**Comment by KevinMenden** [comment]
Hi @kaurravneet4123 , yes there was another change on the modules repo :facepalm: 

`filters.yml` doesn't exist anymore, this is now called `tests/config/pytest_software.yml`

Similarly, the `TOOL/SUBTOOL` template directories have been removed, as there are not needed anymore because of the soon-to-be-released `nf-core modules create` command. 
This should have been addressed during pulling and merging though.
Does that help you?
Sorry again, there is a lot of change happening at `nf-core/modules` lately ....

**Comment by kaurravneet4123** [comment]
> Hi @kaurravneet4123 , yes there was another change on the modules repo 🤦
> 
> `filters.yml` doesn't exist anymore, this is now called `tests/config/pytest_software.yml`
> 
> Similarly, the `TOOL/SUBTOOL` template directories have been removed, as there are not needed anymore because of the soon-to-be-released `nf-core modules create` command.
> This should have been addressed during pulling and merging though.
> Does that help you?
> Sorry again, there is a lot of change happening at `nf-core/modules` lately ....

Thanks for the details @KevinMenden. Yes, I have noticed the changes and the issue is resolved now. Thanks

**Comment by kaurravneet4123** [comment]
Hi @KevinMenden, all pytests successful with cnvkit! 

I just had to create a new bed file (/tests/data/genomics/sarscov2/bed/baits.bed) to tests sarscov2 dataset. Rest all the input files are the same as in the sarscov2 dir.

**Comment by KevinMenden** [comment]
Awesome - getting close now!  The "baits.bed" file is good, might rename this at some point but for now that's 👍 

Could you remove all of the other data you're currently still adding to `tests/data` in your PR? I understand that you only need the sarscov2 data right?

And then the editorconfig of course :) this might be because you need a newline at the end of every file. So easy to fix.
See here for the config:
https://github.com/nf-core/modules/blob/master/.editorconfig

Last two tiny things and then we can merge!

**Comment by kaurravneet4123** [comment]
> Awesome - getting close now! The "baits.bed" file is good, might rename this at some point but for now that's 👍
> 
> Could you remove all of the other data you're currently still adding to `tests/data` in your PR? I understand that you only need the sarscov2 data right?
> 
> And then the editorconfig of course :) this might be because you need a newline at the end of every file. So easy to fix.
> See here for the config:
> https://github.com/nf-core/modules/blob/master/.editorconfig
> 
> Last two tiny things and then we can merge!

Hi @KevinMenden, I have removed the extra data files. Code linting is not becoming successful despite of adding a new line at the end of the files. I have tried it several times.

**Comment by kaurravneet4123** [comment]
Hi @KevinMenden, all 7/7 checks successful eventually :-)

**Review by KevinMenden — APPROVED** [review]

**Comment by KevinMenden** [comment]
Great, looks good to me! I think we can change the name of the `baits.bed` file in a separate PR since it's only used by one tool anyway. So we get this one finally merged :) 

Just need an approval from @drpatelh before merging because he requested changes.

**Comment by kaurravneet4123** [comment]
> Great, looks good to me! I think we can change the name of the `baits.bed` file in a separate PR since it's only used by one tool anyway. So we get this one finally merged :)
> 
> Just need an approval from @drpatelh before merging because he requested changes.

Thanks @KevinMenden. I look forward to getting it approved by @drpatelh :-)

**Comment by kaurravneet4123** [comment]
I have added all the reviewers names (@KevinMenden, @MaxUlysse and @drpatelh) in the authors' list. I hope this is ok with you all.

**Review by MaxUlysse — APPROVED** [review]
LGTM

**Comment by kaurravneet4123** [comment]
> LGTM

Thanks @MaxUlysse :-)

**Comment by kaurravneet4123** [comment]
Hi @drpatelh, I have made all the changes to the module. Is there anything else that needs changing from my side?

File: software/cnvkit/main.nf
Comment (drpatelh) [inline software/cnvkit/main.nf:9]: Is this the appropriate `label` here @kaurravneet4123 ? Or will this process need more resources by default e.g. `process_medium`?

**Review by drpatelh — APPROVED** [review]
Looks great @kaurravneet4123 ! One minor question to address and then feel free to merge.

File: software/cnvkit/main.nf
Comment (kaurravneet4123) [reply software/cnvkit/main.nf:9]: Hi @drpatelh, I think process low if fine as for a wgs data file, it needs only 1.8GB memory. But if you think it is better to change it to 'process_medium', I am happy to do that.

**Comment by kaurravneet4123** [comment]
Hi @drpatelh , sorry for disturbing you so late. I don't have write permission to merge the PR. Would it be possible to do it from your side?

**Comment by kaurravneet4123** [comment]
Thanks @drpatelh and @rivera10

**Pull request merged**
